### PR TITLE
Route Refresh

### LINKED ISSUE #2284: Periodically refresh the route with the latest traffic and ETAs
During turn-by-turn navigation, the SDK should periodically update the current route’s attributes and ETAs to reflect changing traffic conditions.

A specialized Directions API returns a sparse response that we’ll need to merge into the existing route and route progress models. This will require the ability to mutate the route model without recreating it from scratch. Currently most of the properties in Route, RouteLeg, etc. are constant. MapboxDirections.swift could define mutable subclasses, or define these types to structures that come with mutability built-in, or implement built-in methods for merging in updated data.

The route refreshing feature will complement two other forms of updates the SDK already performs: proactive rerouting periodically checks for a more optimal route than the current one, while ETA refreshing simply updates the ETA to reflect whether the user is keeping pace.

This feature has already been implemented on Android in mapbox/mapbox-navigation-android#1855.

/cc @mapbox/navigation-ios

### PR DESCRIPTION
**Pull request by Udumft**
Traffic congestion segments along the route line and the estimated arrival time periodically update to reflect current conditions when using the `DirectionsProfileIdentifier.automobileAvoidingTraffic` profile. These updates correspond to the new `Notification.Name.routeControllerDidRefreshRoute` notification, `NavigationServiceDelegate.navigationService(_:didRefresh:)` method, and `NavigationViewControllerDelegate.navigationViewController(_:didRefresh:)` method. To disable these updates, set `RouteOptions.refreshingEnabled` to `false`.

The `NavigationViewController.route` and `NavigationService.route` properties are now read-only. To change the route that the user is traveling along, set the `NavigationViewController.indexedRoute` or `NavigationService.indexedRoute` property instead, pairing the route with the index of the route in the original `RouteResponse` object.

The following methods now require a route index to be passed in as an argument:

* `NavigationViewController(for:routeIndex:routeOptions:navigationOptions:)`
* `NavigationViewController(route:routeIndex:routeOptions:navigationService:)`
* `CarPlayManagerDelegate.carPlayManager(_:navigationServiceAlong:routeIndex:routeOptions:desiredSimulationMode:)`
* `MapboxNavigationService(route:routeIndex:routeOptions:)`
* `MapboxNavigationService(route:routeIndex:routeOptions:directions:locationSource:eventsManagerType:simulating:routerType:)`
* `RouteProgress(route:routeIndex:options:legIndex:spokenInstructionIndex:)`
* `RouteProgress(route:routeIndex:options:legIndex:spokenInstructionIndex:)`
* `Router(along:routeIndex:options:directions:dataSource:)`
* `RouteController(along:routeIndex:options:directions:dataSource:)`

Resolves #2284 

----

Work-In-Progress
So far added basic implementation for calling Route Refresh. Experimenting with `RouteProgress` to refresh it's Route contents (suggestions are welcome).
Thinking about route refresh period setting to be related with better route search period

### CODE REVIEW

**Comment by asinghal22** [comment]
Discussed today with Antonio and Hossam. Not a release blocker for v1.0 marketing push. We will backport to v1.0 branch and include as part of v1.0.x

**Comment by asinghal22** [comment]
Removed the release blocker sticker.

File: MapboxCoreNavigation/CoreConstants.swift [hunk 1]
```diff
@@ -74,7 +74,7 @@ public var RouteControllerNumberOfSecondsForRerouteFeedback: TimeInterval = 10
 public var RouteControllerMinimumDurationRemainingForProactiveRerouting: TimeInterval = 600
 
 /**
- The number of seconds between attempts to automatically calculate a more optimal route while traveling.
+ The number of seconds between attempts to automatically calculate a more optimal route while traveling. During such attempt, if Route Refresh is enabled, route will also be refreshed to verify correct ETA and congestion.
```
Comment (1ec5) [inline MapboxCoreNavigation/CoreConstants.swift:77]:
```suggestion
 The number of seconds between attempts to automatically calculate a more optimal route while traveling. During such attempt, if Route Refresh is enabled, route will also be refreshed to verify correct ETA and congestion.
```
```suggestion
 The number of seconds between attempts to automatically calculate a more optimal route while traveling.
 
 In addition to calculating a more optimal route, `RouteController` also refreshes time-dependent statistics about the route, such as traffic congestion and the remaining duration, as long as `DirectionsOptions.profileIdentifier` is set to `DirectionsProfileIdentifier.automobileAvoidingTraffic` and `RouteOptions.refreshingEnabled` is set to `true`.
```

File: MapboxCoreNavigation/Router.swift [hunk 2]
```diff
@@ -93,12 +102,68 @@ protocol InternalRouter: class {
...
+                                self?.isRefreshing = false
+                                self?.lastRouteRefresh = nil
+                                completion()
+                            }
+                            
+                            guard case let .success(response) = result else {
+                                return
+                            }
+                            
+                            guard let route = response.route else {
```
Comment (1ec5) [inline MapboxCoreNavigation/Router.swift:160]: As of mapbox/mapbox-directions-swift@3941c256c9938b56ffaea5a8b7443a0d1c24b0e7, `RouteRefreshResponse.route` is merely a skeletal RefreshedRoute, not a fully merged Route. We need to call `Route.refreshLegAttributes(from:)` here.

**Comment by 1ec5** [comment]
I expanded the scope of this PR in #2591. If approved, that PR would merge into this one and we’d be pretty close to landing the feature.

File: MapboxCoreNavigation/Router.swift [hunk 3]
```diff
@@ -93,12 +102,68 @@ protocol InternalRouter: class {
...
+                                return
+                            }
+                            self?.routeProgress.refreshRoute(with: route)
```
Comment (1ec5) [inline MapboxCoreNavigation/Router.swift:163]: This is where we’d typically post a notification and inform the delegate that the route has changed in some way. NavigationViewController would need such a delegate message to know when to redraw the route line with updated traffic congestion segments.

File: MapboxCoreNavigation/CoreConstants.swift (hunk 1 above)
Comment (1ec5) [reply MapboxCoreNavigation/CoreConstants.swift:77]: #2591 will take care of this feedback.

File: MapboxCoreNavigation/Router.swift (hunk 2 above)
Comment (1ec5) [reply MapboxCoreNavigation/Router.swift:160]: #2591 will take care of this feedback.

File: MapboxCoreNavigation/Router.swift (hunk 3 above)
Comment (1ec5) [reply MapboxCoreNavigation/Router.swift:163]: #2591 will take care of this feedback.

**Review by 1ec5 — APPROVED** [review]

**Comment by 1ec5** [comment]
I missed a couple spots when adding `routeIndex:` to various initialization calls: #2598.

**Pull request merged**